# TP 2  -- éléments de correction

*Renommez ce notebook en y ajoutant vos initiales (par exemple `TP2_PP.ipynb`), **sans espace**, et pensez à le sauvegarder régulièrement !*

***N’ajoutez pas et ne supprimez aucune cellule.***

*Renseignez ci-dessous vos prénoms et noms (double-cliquez pour éditer la cellule).*

---

Prénom : 

Nom : 

---

## Partie 1 : Calcul du %GC pour le génome du Sars-Cov-2

Le génome de référence du Sars-Cov-2 est disposnible sous l'identifiant [NC_045512](https://www.ncbi.nlm.nih.gov/nuccore/NC_045512) sur le site du NCBI. Il est constitué de 29903 bases.

Le fichier `sars-cov-2.txt` contient le génome de ce virus. Toute la séquence se trouve sur une seule ligne.

Pour :

- lire ce fichier, 
- extraire la séquence du génome de Sars-Cov-2
- et la stocker dans la variable `sequence`, 

vous pouvez utiliser le code suivant :

In [1]:
with open("sars-cov-2.txt", "r") as txt_file:
    sequence = txt_file.read()

Affichez la taille de la séquence lue dans la variable `sequence` :

In [2]:
print(len(sequence))

29903


Affichez les 10 premières bases de la séquence :

(*Vous devriez obtenir `ATTAAAGGTT`*)

In [3]:
print(sequence[0:10])

ATTAAAGGTT


Affichez les 10 dernières bases de la séquence :

(*Vous devriez obtenir `AAAAAAAAAA`*)

In [4]:
print(sequence[-10:])

AAAAAAAAAA


Réutilisez le code de la partie 3 du TP1, pour calculer le %GC de cette sequence :

In [5]:
nb_A = 0
nb_T = 0
nb_C = 0
nb_G = 0
for base in sequence:
    if base == "A":
        nb_A = nb_A + 1
    elif base == "T":
        nb_T = nb_T + 1
    elif base == "C":
        nb_C = nb_C + 1
    elif base == "G":
        nb_G = nb_G + 1

GC = (nb_G + nb_C)/(nb_A + nb_T + nb_G + nb_C) * 100

Affichez la valeur du %GC avec **deux chiffres** après la virgule, sous la forme :

```
Le pourcentage de GC est : xx.xx
```

(*Respectez scrupuleusement le format d'affichage demandé.*)

In [6]:
print(f"Le pourcentage de GC est : {GC:.2f}")

Le pourcentage de GC est : 37.97


## Partie 2 : Calcul du %GC pour le génome du Sars-Cov-2 à partir d'un fichier au format FASTA

### Format FASTA

En biologie, les séquences ne se trouvent jamais sous la forme d'une séquence par ligne.

Un format classique pour partager des séquences est le format FASTA :
```
>en-tête
séquence avec un nombre maximum de caractères par ligne
séquence avec un nombre maximum de caractères par ligne
séquence avec un nombre maximum de caractères par ligne
séquence avec un nombre maximum de caractères par ligne
séquence avec un nombre max
```

Dans le cas du génome du Sars-cov-2 stocké au format FASTA dans le fichier `sars-cov-2.fasta`, les premières lignes sont :

```
>NC_045512.2 Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome
ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGTAGATCTGTTCTCTAAA
CGAACTTTAAAATCTGTGTGGCTGTCACTCGGCTGCATGCTTAGTGCACTCACGCAGTATAATTAATAAC
TAATTACTGTCGTTGACAGGACACGAGTAACTCGTCTATCTTCTGCAGGCTGCTTACGGTTTCGTCCGTG
TTGCAGCCGATCATCAGCACATCTAGGTTTCGTCCGGGTGTGACCGAAAGGTAAGATGGAGAGCCTTGTC
CCTGGTTTCAACGAGAAAACACACGTCCAACTCAGTTTGCCTGTTTTACAGGTTCGCGACGTGCTCGTAC
GTGGCTTTGGAGACTCCGTGGAGGAGGTCTTATCAGAGGCACGTCAACATCTTAAAGATGGCACTTGTGG
CTTAGTAGAAGTTGAAAAAGGCGTTTTGCCTCAACTTGAACAGCCCTATGTGTTCATCAAACGTTCGGAT
```

Sachant que :

- la séquence du Sars-cov-2 contient 29903 bases
- et que le `sars-cov-2.fasta` contient 70 bases par lignes,

déterminez avec des variables et des opérateurs Python combien de lignes devrait contenir le fichier `sars-cov-2.fasta` de la séquence du génome du Sars-cov-2 au format FASTA. N'oubliez pas de compter la ligne d'entête.

(*Pour répondre à cette question, ne parcourez pas le fichier avec Python.*)

Indiquez et justifiez votre réponse dans les deux cellules ci-dessous :

In [7]:
# Indiquez dans cette cellule votre calcul avec du code Python.

(29903 / 70) + 1

428.1857142857143

*Double-cliquez sur cette cellule pour justifier ici votre réponse.* 

Nombre de lignes total = nombre de lignes de séquence + ligne entête.

On arrondit 428.1857142857143 au nombre entier supérieur, soit 429 lignes

Ouvrez le fichier `sars-cov-2.fasta` en double-cliquant sur le nom du fichier dans le navigateur de fichiers à gauche.

Vérifiez que le nombre de lignes que vous avez calculé correspond bien au nombre de lignes dans le fichier.

Remarque : le numéro de la ligne est affiché en gris à gauche de chaque ligne.

*Double-cliquez sur cette cellule pour discuter ici votre analyse
et comparer le nombre de lignes du fichier obtenu avec Python et celui obtenu en ouvrant le fichier.*

Le fichier sars-cov-2.fasta possède 429 lignes ce qui correspond bien au nombre de lignes attendu.

### Lecture d'un séquence au format FASTA

Pour lire une séquence au format FASTA, il faut se souvenir :
- que la première ligne contient l'entête (l'identifiant + le nom + d'autres informations) ;
- que la séquence est répartie sur les lignes suivantes.

Voici un exemple de code qui vous pouvez utiliser pour lire la séquence depuis un fichier au format FASTA :

In [8]:
with open("sars-cov-2.fasta", "r") as fasta_file:
    header = ""
    sequence = ""
    for line in fasta_file:
        if line.startswith(">"):
            header = line.strip()
        else:
            sequence += line.strip()

L'entête est stocké dans la variable `header`. La séquence est stockée dans la variable `sequence`.

Prenez quelques minutes pour comprendre le fonctionnement de ce code Python.

Affichez ensuite l'entête :

In [9]:
print(header)

>NC_045512.2 Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome


Affichez la taille de la séquence :

In [10]:
print(len(sequence))

29903


Affichez les 10 premières bases de la séquence :

(*Vous devriez obtenir `ATTAAAGGTT`*)

In [11]:
print(sequence[0:10])

ATTAAAGGTT


Affichez les 10 dernières bases de la séquence :

(*Vous devriez obtenir `AAAAAAAAAA`*)

In [12]:
print(sequence[-10:])

AAAAAAAAAA


Réutilisez le code de la partie 1 de ce TP, pour calculer le %GC de cette sequence :

In [13]:
nb_A = 0
nb_T = 0
nb_C = 0
nb_G = 0
for base in sequence:
    if base == "A":
        nb_A = nb_A + 1
    elif base == "T":
        nb_T = nb_T + 1
    elif base == "C":
        nb_C = nb_C + 1
    elif base == "G":
        nb_G = nb_G + 1

GC = (nb_G + nb_C)/(nb_A + nb_T + nb_G + nb_C) * 100

Affichez la valeur du %GC avec deux chiffres après la virgule, sous la forme :
```
Le pourcentage de GC est : xx.xx
```

(*Respectez scrupuleusement le format d'affichage demandé.*)

In [14]:
print(f"Le pourcentage de GC est : {GC:.2f}")

Le pourcentage de GC est : 37.97


Obtenez-vous le même résultat que dans la partie 1 ?

*Double-cliquez sur cette cellule pour justifier ici votre réponse.* 

Oui, on obtient le même résultat.

Le génome est le même, mais présenté dans des formats différents.

## Partie 3 : Comparaison de génomes

La même procédure peut être utilisée pour étudier le génome d'un autre coronavirus : le [MERS-CoV](https://www.ncbi.nlm.nih.gov/nuccore/NC_019843).

Réutilisez et adaptez le code précédant pour effectuer les étapes suivantes :

1. Lire le fichier `mers-cov.fasta` et extraire la séquence du génome du MERS-Cov.
2. Afficher la taille du génome et vérifier qu'elle correspond à celle indiquée sur [la page du NCBI](https://www.ncbi.nlm.nih.gov/nuccore/NC_019843).
3. Calculer le %GC de ce génome.
4. Afficher la valeur du %GC avec deux chiffres après la virgule, sous la forme :
```
Le pourcentage de GC est : xx.xx
```

(*Respectez scrupuleusement le format d'affichage demandé.*)

### Étape 1

In [15]:
with open("mers-cov.fasta", "r") as fasta_file:
    header = ""
    sequence = ""
    for line in fasta_file:
        if line.startswith(">"):
            header = line.strip()
        else:
            sequence += line.strip()

### Étape 2

In [16]:
print(len(sequence))

# La taille obtenue est la même que celle fournie sur le site du NCBI :
# LOCUS       NC_019843              30119 bp    RNA     linear   VRL 08-AUG-2020

30119


### Étape 3

In [17]:
nb_A = 0
nb_T = 0
nb_C = 0
nb_G = 0
for base in sequence:
    if base == "A":
        nb_A = nb_A + 1
    elif base == "T":
        nb_T = nb_T + 1
    elif base == "C":
        nb_C = nb_C + 1
    elif base == "G":
        nb_G = nb_G + 1

GC = (nb_G + nb_C)/(nb_A + nb_T + nb_G + nb_C) * 100

### Étape 4

In [18]:
print(f"Le pourcentage de GC est : {GC:.2f}")

Le pourcentage de GC est : 41.24


## Pour terminer avec un peu d'humour

![](https://imgs.xkcd.com/comics/coronavirus_genome.png)

Source : [XKCD](https://xkcd.com/2298/), licence Creative Commons BY-NC